homeassistant/components/folder_watcher.py

```diff
@@ -0,0 +1,118 @@
+"""
+Component for monitoring activity on a folder.
+
+For more details about this platform, refer to the documentation at
+https://home-assistant.io/components/folder_watcher/
+"""
+import os
+import logging
+import voluptuous as vol
+from homeassistant.helpers.entity import Entity
+from homeassistant.const import (
+    EVENT_HOMEASSISTANT_START, EVENT_HOMEASSISTANT_STOP)
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['watchdog==0.8.3']
+_LOGGER = logging.getLogger(__name__)
+
+CONF_FOLDER = 'folder'
+CONF_PATTERNS = 'patterns'
+CONF_WATCHERS = 'watchers'
+DEFAULT_PATTERN = '*'
+DOMAIN = "folder_watcher"
+EVENT_TYPE = "event_type"
+FILE = 'file'
+FOLDER = 'folder'
+
+WATCHER_CONFIG_SCHEMA = vol.Schema([{
+    vol.Required(CONF_FOLDER): cv.isdir,
+    vol.Optional(CONF_PATTERNS, default=[DEFAULT_PATTERN]): vol.All(
+        cv.ensure_list, [cv.string]),
+}])
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_WATCHERS): WATCHER_CONFIG_SCHEMA,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up the folder watcher."""
+    conf = config[DOMAIN]
+    watchers = conf[CONF_WATCHERS]
+
+    def run_setup(event):
+        """"Wait for HA start then setup."""
+        for watcher in watchers:
+            path = watcher[CONF_FOLDER]
+            patterns = watcher[CONF_PATTERNS]
+            if not hass.config.is_allowed_path(path):
+                _LOGGER.error("folder %s is not valid or allowed", path)
+                continue
+            else:
+                Watcher(path, patterns, hass)
+
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, run_setup)
```

Comment:
I'm not sure if this is necessarily the correct way. Now we only add the sensors when Home Assistant is started, but I think we should only *register the observer* when Home Assistant is started. I could very well be wrong about this, but I think we should initialize entities at `setup` time, and not at `EVENT_HOMEASSISTANT_START` time.

```python
# in setup()
conf = config[DOMAIN]
watchers = conf[CONF_WATCHERS]
for watcher in watchers:
    path = watcher[CONF_FOLDER]
    patterns = watcher[CONF_PATTERNS]
    if not hass.config.is_allowed_path(path):
        _LOGGER.error("folder %s is not valid or allowed", path)
        continue
    Watcher(path, patterns, hass)
```

```python
# in Watcher#__init__
self._observer = Observer()
hass.bus.listen_once(EVENT_HOMEASSISTANT_START, self.startup)
hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, self.shutdown)
```

```python
# in Watcher#startup
self._observer.schedule(
    create_event_handler(patterns, hass),
    path,
    recursive=True)
self._observer.start()
```

Reply:
@OttoWinter I've tried that out but `self._observer.start()` is never called - presumably because `Watcher#__init__` is called by `EVENT_HOMEASSISTANT_START` and `self.startup` is listening for another `EVENT_HOMEASSISTANT_START` which will never come